plugin/src/com/microsoft/alm/plugin/external/visualstudio/VisualStudioTfvcClient.java

```diff
@@ -0,0 +1,110 @@
+    private static final String EDITION_RELATIVE_TF_EXE_PATH =
+            "Common7\\IDE\\CommonExtensions\\Microsoft\\TeamFoundation\\Team Explorer\\TF.exe";
+
+    private static final ToolVersion MINIMAL_SUPPORTED_VERSION = new ToolVersion(15, 0, 0, "");
+
+    @Nullable
+    public static Path getOrDetectPath(PropertyService propertyService) {
+        String path = propertyService.getProperty(PropertyService.PROP_VISUAL_STUDIO_TF_CLIENT_PATH);
+        if (!StringUtil.isEmpty(path)) {
+            return Paths.get(path);
+        }
+
+        return detectClientPath();
+    }
+
+    /**
+     * Tries to detect path to the latest available TFVC client from a set of hardcoded locations.
+     *
+     * @return detected path or null.
+     */
+    @Nullable
+    public static Path detectClientPath() {
+        if (!SystemInfo.isWindows) return null;
+        String programFilesPath = SystemInfo.is32Bit ? System.getenv("ProgramFiles") : System.getenv("ProgramFiles(x86)");
+        if (StringUtils.isEmpty(programFilesPath))
+            programFilesPath = "C:\\Program Files (x86)";
+        Path visualStudioPath = Paths.get(programFilesPath, "Microsoft Visual Studio");
+        for (String vsVersion : SUPPORTED_VS_VERSIONS) {
+            Path versionPath = visualStudioPath.resolve(vsVersion);
+            if (!versionPath.toFile().isDirectory()) continue;
+            for (String vsEdition : SUPPORTED_VS_EDITIONS) {
+                Path editionPath = versionPath.resolve(vsEdition);
+                Path tfExePath = editionPath.resolve(EDITION_RELATIVE_TF_EXE_PATH);
+                if (tfExePath.toFile().isFile()) return tfExePath;
+            }
+        }
+
+        return null;
+    }
+
+    @NotNull
+    public static CompletionStage<Void> checkVersionAsync(Path visualStudioClientPath) {
+        return VisualStudioTfvcCommands.getVersionAsync(visualStudioClientPath).thenAccept(version -> {
+            if (version == null)
+                throw new VisualStudioClientVersionException(ToolVersion.UNKNOWN, MINIMAL_SUPPORTED_VERSION);
+
+            if (MINIMAL_SUPPORTED_VERSION.compare(version) > 0)
+                throw new VisualStudioClientVersionException(version, MINIMAL_SUPPORTED_VERSION);
+        });
+    }
+
+    static List<String> executeClientAndGetOutput(
+            @NotNull Path clientPath,
+            @Nullable Path workingDirectory,
+            @NotNull List<String> arguments) throws IOException, InterruptedException {
+        ourLogger.info("Executing VS client: " + clientPath + ", args: " + StringUtils.join(arguments, ','));
+        List<String> command = new ArrayList<>(arguments.size() + 1);
+        command.add(clientPath.toString());
+        command.addAll(arguments);
+
+        Process client = new ProcessBuilder()
+                .command(command)
+                .directory(workingDirectory == null ? null : workingDirectory.toFile())
```

Comment:
Maybe we should use project dir as default when null passed.

Reply:
Agreed. I've modified the defaults around here, but the parameter will still be nullable since `Project::getBasePath()` is marked as nullable. There're cases (such as `VersionCommand`) when the path isn't very important (but we'll still try to run these commands in the project base directory if we can).